### DISCUSSION TITLE
how to read environment variables in javascript when extended with pocketbase

### DISCUSSION THREAD

**Discussion by nikhil-varma-mudunuri** [pocketbase/pocketbase#3181] [Q&A] [answered]
same as title

**Comment by ganigeorgiev** [comment, marked as answer]
The `process` module is loaded so you should be able to access env variables similar as in Node.js, aka. `process.env.MY_VAR`.

Alternatively, you can also use [`$os.getenv("MY_VAR")`](https://pocketbase.io/jsvm/functions/_os.getenv.html).

**Reply by sneakylenny** [reply]
I would love to see this documented more explicitly. The only mention I could find was in a comment on a code example here: https://pocketbase.io/docs/js-migrations/#creating-initial-superuser

**Comment by nikhil-varma-mudunuri** [comment]
how can i read data in .env file and use in code , this code `process.env.MY_VAR` is not working for .env file

**Reply by ganigeorgiev** [reply]
The `.env` file has no special meaning in PocketBase (_and I think this is also the case for Node, at least for v20.4.0_).

`process.env.MY_VAR` returns the OS `MY_VAR` env variable (_if it is defined for your current user_).

If you want to read "ENV variables" from an `.env` file  you'll have to manually read it and parse its content on your own. I personally would avoid doing this and instead define the variables in a `.json` file and require it as a regular js module, for example:

```js
// pb_hooks/config.json
{
    something: "test"
}
```

```js
// pb_hooks/main.pb.js

// (<= v0.22.x) after application bootstrap, or any other hook...
onAfterBootstrap((e) => {
    const config = require(`${__hooks}/config.json`)

    console.log(config.something)
})

// (v0.23+) after application bootstrap, or any other hook...
onBootstrap((e) => {
    e.next()

    const config = require(`${__hooks}/config.json`)

    console.log(config.something)
})
```

This also has the benefit that the config values could be numbers, arrays, objects, etc., not just plain strings.

**Reply by groenroos** [reply]
While a `.json` file would be easier out of the box, all my other stuff reads environment variables from `.env`, and I don't want to worry about storing/syncing variables in multiple places in the repo.

I wrote a little `env.js` like so, copying the gist of the `parse()` method in the `dotenv` npm package (since I couldn't figure out a way to load `dotenv` directly):

```javascript
/* Definition of a line, copied from dotenv */
const LINE = /(?:^|^)\s*(?:export\s+)?([\w.-]+)(?:\s*=\s*?|:\s+?)(\s*'(?:\\'|[^'])*'|\s*"(?:\\"|[^"])*"|\s*`(?:\\`|[^`])*`|[^#\r\n]+)?\s*(?:#.*)?(?:$|$)/mg

/* Read env file */
const envFile = $os.readFile('.env');
const envStr = String.fromCharCode(...envFile).replace(/\r\n?/mg, '\n');

/* Go through each line */
let match;
while ((match = LINE.exec(envStr)) != null) {
	const key = match[1];
	let value = (match[2] || '').trim();

	/* Check if double quoted */
	const maybeQuote = value[0];

	/* Remove surrounding quotes */
	value = value.replace(/^(['"`])([\s\S]*)\1$/mg, '$2');

	/* Expand newlines if double quoted */
	if (maybeQuote === '"') {
		value = value.replace(/\\n/g, '\n');
		value = value.replace(/\\r/g, '\r');
	}

	process.env[key] = value;
}
```

Then, in each `pb_hooks` file where I need my environment variables, a simple `require` is enough to populate `process.env`:

```
require('./env.js');
console.log(process.env.SOME_VAR);
```

**Reply by ersinakinci** [reply]
I slightly modified @groenroos's script to first check for whether `.env` exists so that it fails gracefully:

```js
/* Definition of a line, copied from dotenv */
const LINE =
  /(?:^|^)\s*(?:export\s+)?([\w.-]+)(?:\s*=\s*?|:\s+?)(\s*'(?:\\'|[^'])*'|\s*"(?:\\"|[^"])*"|\s*`(?:\\`|[^`])*`|[^#\r\n]+)?\s*(?:#.*)?(?:$|$)/gm;
let envFile, envStr;

/* Read env file */
try {
  envFile = $os.readFile(".env");
  envStr = String.fromCharCode(...envFile).replace(/\r\n?/gm, "\n");
} catch (e) {
  // No .env file, noop
}

if (envFile && envStr) {
  /* Go through each line */
  let match;
  while ((match = LINE.exec(envStr)) != null) {
    const key = match[1];
    let value = (match[2] || "").trim();

    /* Check if double quoted */
    const maybeQuote = value[0];

    /* Remove surrounding quotes */
    value = value.replace(/^(['"`])([\s\S]*)\1$/gm, "$2");

    /* Expand newlines if double quoted */
    if (maybeQuote === '"') {
      value = value.replace(/\\n/g, "\n");
      value = value.replace(/\\r/g, "\r");
    }

    process.env[key] = value;
  }
}
```

I saved it in `pb_hooks/lib/env.js`, and then I load it when the app starts:

```js
// pb_hooks/main.pb.js

// Read the .env file and set the environment variables.
onAfterBootstrap((e) => {
  require(`${__hooks}/lib/env.js`);
});
```

**Reply by aymericdo** [reply]
It's working well on my machine, but when I try to send it in production through docker, the path is not working...

**GoError: open .env: no such file or directory**

**Reply by theodorewilkins** [reply]
I'd recommend using process.env.MY_VAR as suggested by @ganigeorgiev. I experimented with @groenroos’s env.js approach, but found it slower and encountered issues with variables not being parsed quickly enough. Using process.env is faster and more reliable, as variables are immediately available at runtime.
